src/material-experimental/mdc-snack-bar/snack-bar-container.ts

```diff
@@ -0,0 +1,174 @@
+  host: {
+    '[attr.role]': '_role',
+    '[class.mdc-snackbar]': 'true',
+    '[class.mat-mdc-snack-bar-container]': 'true',
+    '[class.mat-snack-bar-container]': 'false',
+  }
+})
+export class MatSnackBarContainer extends BasePortalOutlet
+    implements MatSnackBarContainerInterface, AfterViewChecked, OnDestroy {
+  /** Subject for notifying that the snack bar has exited from view. */
+  readonly _onExit: Subject<any> = new Subject();
+
+  /** Subject for notifying that the snack bar has finished entering the view. */
+  readonly _onEnter: Subject<any> = new Subject();
+
+  /** ARIA role for the snack bar container. */
+  _role: 'alert' | 'status' | null;
+
+  private _mdcAdapter: MDCSnackbarAdapter = {
+    addClass: (className: string) => this._setClass(className, true),
+    removeClass: (className: string) => this._setClass(className, false),
+    announce: () => {},
+    notifyClosed: () => this._onExit.next(),
+    notifyClosing: () => {},
+    notifyOpened: () => this._onEnter.next(),
+    notifyOpening: () => {},
+  };
+
+  private _mdcFoundation = new MDCSnackbarFoundation(this._mdcAdapter);
+
+  /** The portal outlet inside of this container into which the snack bar content will be loaded. */
+  @ViewChild(CdkPortalOutlet, {static: true}) _portalOutlet: CdkPortalOutlet;
+
+  /** Element that acts as the MDC surface container which should contain the label and actions. */
+  @ViewChild('surface', {static: true}) _surface: ElementRef;
+
+  /**
+   * Element that will have the `mdc-snackbar__label` class applied if the attached component
+   * or template does not have it. This ensures that the appropriate structure, typography, and
+   * color is applied to the attached view.
+   */
+  @ViewChild('label', {static: true}) _label: ElementRef;
+
+  constructor(
+      private _elementRef: ElementRef<HTMLElement>,
+      public snackBarConfig: MatSnackBarConfig) {
+    super();
+
+    // Based on the ARIA spec, `alert` and `status` roles have an
+    // implicit `assertive` and `polite` politeness respectively.
+    if (snackBarConfig.politeness === 'assertive' && !snackBarConfig.announcementMessage) {
+      this._role = 'alert';
+    } else if (snackBarConfig.politeness === 'off') {
+      this._role = null;
+    } else {
+      this._role = 'status';
+    }
+
+    // `MatSnackBar` will use the config's timeout to determine when the snack bar should be closed.
+    // Set this to `-1` to mark it as indefinitely open so that MDC does not close itself.
+    this._mdcFoundation.setTimeoutMs(-1);
+  }
+
+  ngAfterViewChecked() {
+    // Check to see if the attached component or template uses the MDC template structure,
+    // specifically the MDC label. If not, the container should apply the MDC label class to this
+    // component's label container, which will apply MDC's label styles to the attached view.
+    const attachedViewHasMdcLabel =
+        !!this._label.nativeElement.querySelector('.mdc-snackbar__label');
```

Comment:
This is being run on every change detection cycle so it's very performance sensitive. Can we avoid query the DOM and deduce the class some other way?

Reply:
This is my least favorite part of the implementation, but I'm not sure how to get around it. As far as I can tell, it'll be difficult for the container to know whether the user provided a label and action using the MDC classes (either directly or with our directives).

I'd like to do this once, but I suspect we'll see bugs where the user provides something wrapped in an `ngIf` or `ng-template` that is loaded dynamically. Perhaps there could be a compromise where we perform this once at initialization, and allow for an API that lets the user explicitly say that their content will dynamically have the MDC classes

Reply:
Maybe we can match it with a directive and query it with a `ContentChild`?

Reply:
That seems like it would work

Reply:
The user-provided component or template doesn't get included as part of either the content or view of the snack bar container. Technically I think the portal could provide this info through its view container - is it possible to make the query on the container's behalf?

Reply:
Ah, I see, I forgot it was in a component (my brain spaced on this being snackbar). 

Since snackbar content is generally simple and because they're short-lived, this is probably fine.